kbmd/door.c

```diff
@@ -0,0 +1,483 @@
+
+	/*
+	 * We can't easily log the whole error chain, so we just log the
+	 * topmost one -- however the client gets the entire error chain.
+	 */
+	(void) bunyan_debug(tlog, "Returning error",
+	    BUNYAN_T_STRING, "err_name", errf_name(ef),
+	    BUNYAN_T_STRING, "err_msg", errf_message(ef),
+	    BUNYAN_T_STRING, "err_func", errf_function(ef),
+	    BUNYAN_T_STRING, "err_file", errf_file(ef),
+	    BUNYAN_T_UINT32, "err_line", errf_line(ef),
+	    BUNYAN_T_END);
+
+	errf_free(ef);
+	kbmd_ret_nvlist(nvret);
+}
+
+/*
+ * Return function used by the various commands. If no errors are present,
+ * we return sucess with the given response (which may be NULL if there is
+ * no data to return to the client). If any errors are present, we ignore
+ * the contents of resp and send back the error. The assumption is that
+ * if resp is non-NULL it at best contains a partially constructed response
+ * that failed somewhere mid-way through handling the request. That is in
+ * the case of an error, we don't send any additional data beyond the
+ * data in errval.
+ *
+ * In all cases, both errval and resp are freed prior to calling door_return.
+ */
+void __NORETURN
+kbmd_return(errf_t *restrict errval, nvlist_t *restrict resp)
+{
+	if (errval != ERRF_OK) {
+		nvlist_free(resp);
+		kbmd_ret_error(errval);
+	}
+
+	kbmd_ret_nvlist(resp);
+}
+
+static void
+kbmd_door_server(void *cookie, char *argp, size_t arg_size, door_desc_t *dp,
+    uint_t n_desc)
+{
+	errf_t *ret = ERRF_OK;
+	nvlist_t *req = NULL;
+	tdata_t *td = tdata_get();
+
+	/*
+	 * If we make the kbmd door server use a private thread pool,
+	 * each thread should gets it's own preallocated tdata_t and
+	 * we can remove this check.
+	 */
+	if (td == NULL) {
+		VERIFY0(door_return(generr, generr_sz, NULL, 0));
+	}
+
+	if (!tdata_init(td, dp, n_desc)) {
+		VERIFY0(door_return(generr, generr_sz, NULL, 0));
+	}
+
+	ret = envlist_unpack(argp, arg_size, &req);
+	if (ret != ERRF_OK) {
+		(void) bunyan_info(tlog, "Unable to unpack request",
+		    BUNYAN_T_INT32, "errno", errf_errno(ret),
+		    BUNYAN_T_STRING, "errmsg", errf_message(ret),
+		    BUNYAN_T_END);
+		errf_free(ret);
+		VERIFY0(door_return(generr, generr_sz, NULL, 0));
+	}
+
+	dispatch_request(req);
+}
+
+static void
+create_generr(void)
+{
+	nvlist_t *nvl;
+
+	VERIFY0(nvlist_alloc(&nvl, NV_UNIQUE_NAME, 0));
+	fnvlist_add_boolean_value(nvl, KBM_NV_SUCCESS, B_FALSE);
+	fnvlist_add_string(nvl, KBM_NV_ERRMSG, "general error");
+	VERIFY0(nvlist_pack(nvl, &generr, &generr_sz, NV_ENCODE_NATIVE, 0));
+}
+
+int
+kbmd_door_setup(const char *path)
+{
+	int ret, fd;
+
+	create_generr();
+
+	/*
+	 * XXX: We might want to manually control door server thread creation
+	 * to control memory and resource usage for requests.  More likely to
+	 * be of concern once we add per-zone token support.
+	 */
+	door_fd = door_create(kbmd_door_server, NULL, 0);
+	if (door_fd == -1)
+		return (errno);
+
+	if ((fd = open(path, O_CREAT|O_RDWR, 0666)) == -1) {
```

Comment:
Originally I was thinking we could use authorizations to control access... though everything in illumos still uses solaris.* -- I don't know if there's been any discussion of just continuing to use that, or start a 'illumos' hierarchy (I don't believe any new ones have been added since the split). We can probably defer that and leave it root only for now.

Reply:
(and update perms on the door file accordingly)

Reply:
confirmed.